### ISSUE TITLE
OS Compute (Nova) [1423648]: race conditions with server group scheduler policies

### ISSUE THREAD

**Issue by b3rnard0** [rcbops/rpc-openstack#1144]
##### What version (commit) do you encounter this issue on?

TBD

##### What type of issue is this (bug, feature, documentation, regression)?

bug

##### What is the expected behavior?

See https://bugs.launchpad.net/nova/+bug/1423648

##### What is actually happening?

See https://bugs.launchpad.net/nova/+bug/1423648

##### What were you attempting to do when you encountered this error (give as much detail as possible)? 

See https://bugs.launchpad.net/nova/+bug/1423648

##### How can we reproduce this issue?

See https://bugs.launchpad.net/nova/+bug/1423648

##### How can we verify your issue is addressed?

See https://bugs.launchpad.net/nova/+bug/1423648

##### If this is a large issue, what can be done immediately to alleviate the issue while working on a longer-term solution?

See https://bugs.launchpad.net/nova/+bug/1423648

##### What is your evaluation on the importance of this issue, and why?

See https://bugs.launchpad.net/nova/+bug/1423648

----------------------------------------------------------------------------------------------------------------------------
###### For each pull request, update the branch line to reference it.
- [ ] master
- [ ] liberty-12.2
- [ ] kilo

##### Upgrade impact. Consider the impact on moving between major, minor and patch releases.
- [ ] This change does not impact upgrades
- [ ] This change addresses a potential upgrade impact
- [ ] The upgrade impact is being tracked by another issue.

The issue tracking the upgrade impact is:

##### Testing impact
- [ ] An issue/pull-request has been create or exists to track the testing gap, and is referenced below
- [ ] This change includes additional testing
- [ ] This change is tested by the existing setup

Additional testing links:

##### Release note
- [ ] This change includes a release note.
- [ ] This change does not require a release note.

##### Does your change require documentation?
- [ ] Yes, and I submitted an issue to [Docs issues](https://github.com/rackerlabs/docs-rpc/issues "Docs issues")
- [ ] Yes, and I contributed a documentation patch to [docs-rpc](https://github.com/rackerlabs/docs-rpc "docs-rpc")
- [ ] No documentation required for this change

Additional documentation links (please add links if you have ticked yes):

**Comment by b3rnard0** [comment]
We will need more details so that we can start tracking this issue. cc: @BjoernT @git-harry

**Comment by git-harry** [comment]
Marked as blocked because the upstream issue is currently being worked by a member of the community.

**Comment by git-harry** [comment]
It has been requested that we take a more proactive approach with this issue so I have removed the blocked label.

**Comment by alextricity25** [comment]
UPDATE: I have attempted to get a progress report from the community member working the bug; no response yet.
In the mean time, I'm working on reproducing the issue. Due to the nature of the bug, a multi-node needs to be brought up. I'm finishing that up today, and will continue my investigation tomorrow.

**Comment by alextricity25** [comment]
UPDATE: The community member has dropped the bug, so I assigned it to myself.
I'm not too familiar with the nova codebase, but I can definitely learn ^.^

I don't think this will be a big change, I just need some time to learn the code base and find the best solution.

**Comment by alextricity25** [comment]
UPDATE: I haven't been able to reproduce the issue yet. Good news is that I have an environment up that *should* be suitable for reproducing this. I also have some deeper insight as to what the solution might be, I just need to figure out where in the nova codebase would be a good place for the logic. I've contacted one of our nova guys in Australia to see if he could help out a bit.

**Comment by b3rnard0** [comment]
@git-harry We need feedback from a Nova contributor. @alextricity25 will try to reach out to someone in the AUS team to see if we can get support for this issue.

**Comment by alextricity25** [comment]
UPDATE: I was able to meet with @tbreeds to get help with some of the nova code. There is still a lot to work though, and it's looking like we are going to need some prerequisite patches before the solution to this problems goes through. I'm keeping this as blocked until we finish working through these patches.

**Comment by git-harry** [comment]
This is currently blocked due to lack of assignee, @alextricity25 has been moved onto other tasks. This can be unblocked as someone becomes available.

**Comment by tomc603** [comment]
@alextricity25 Would you mind writing up a bit of a hand-off on this so we don't duplicate work? You mentioned that you have some insight into what the solution might be and you mentioned speaking with @tbreeds. Do you have any notes you could put here to help us pick up where you left off?

**Comment by alextricity25** [comment]
I started doing work based off an [abandoned patch](https://review.openstack.org/#/c/164762/9) that stalled a bit. I took the same solution, but moved some of the code around to accommodate what was said in the comments of said patch.
@tbreeds noticed some flaws with the testing of test_get_hosts(), so that needed to be fixed before I could even get the unit tests to pass. As far as I know, those changes aren't in upstream. Much less upstream liberty. I'll send you an email thread with some patch formats that you can apply if you want to try them out.

Also, I was never able to reproduce the issue. I tried creating 10 instances on 2 compute hosts on the same server group with the affinity policy(multiple times), but I never encountered the race condition.

**Comment by tbreeds** [comment]
Yeah getting this fixed in liberty is going to be nearly impossible.  Given it will EOL in ~10weeks and we don't have a fix in master yet.

Focus on master and then we'll see where we wash up.  If I haven't done so already I'll push the testing fixes once Ocata opens up

**Comment by tomc603** [comment]
@rcbops/support-feedback Who would be the best person on the support team to discuss this issue with? I have some important feedback to give, and I need to make sure the proper eyes get on it.

**Comment by BjoernT** [comment]
@tomc603 Why not updating this thread, this issue was originally reported by @JCallicoat I believe

**Comment by tomc603** [comment]
My feedback is long and winding, but I'll gladly put it here for posterity.

This bug seems to have been discovered quite a while ago (Kilo is tagged in the initial report), so I'm not entirely sure of the full history or the severity. I'd like to get more details about the impact to customers/support, and the frequency we see this issue crop up. Obviously backporting to Kilo isn't a possibility, and we do not believe backporting to Liberty will be accepted by upstream Nova at this point. But, we can certainly massage the existing backport and the work @alextricity25 did to try to get some of the issue addressed for Mitaka.

Having said that, this patch won't actually address the race condition. It may shrink the time window in which the race can be triggered, but fundamentally the condition will still be present. @tbreeds mentioned the fact that in the O release development time, a new scheduling system is being worked on which is due to drop in the P release. This new scheduler changes the way resources are scheduled from the ground up, and should actually be addressing these kinds of corner cases from a more correct approach. But obviously we don't want to wait over a year to release this to our customers.

Alternatively, a work-around (which actually sounds like the more correct way to perform these near simultaneous boots) is to do a nova multi-instance boot. When doing a multi-instance boot, the scheduler collects all of the necessary affinity information in a more correct manner and uses that affinity information appropriately when scheduling instances onto hosts.

Given the existence of multi-instance boot, I'd like to know if it's even possible for us to educate customers and/or support folks about when it's appropriate and in fact more safe to use that over a rapid succession boot of single instances. This is obviously a much simpler approach and avoids the race condition entirely, which again will exist even after this patch is merged upstream.

Anyway, I just need to know more about the situation surrounding this bug so we can work together to best address it. Feedback from anybody in support that has encountered this bug, including the sequence of events that led to this bug being encountered, would really be appreciated.

**Comment by tomc603** [comment]
And I forgot to tag @rcbops/support-feedback @BjoernT and @JCallicoat in the above message. Great.

**Comment by TEKNISHEN** [comment]
@tomc603 - I've pinged Jordan as well on this.

**Comment by tomc603** [comment]
@rcbops/support-feedback I still haven't heard anything about this, and I have other priority work coming up soon. I'd like to get this resolved for everybody before that work becomes an emergency.

**Comment by JCallicoat** [comment]
@tomc603 I don't know of this being a huge pain point for customers. I only know of a handful using affinity groups and haven't heard of this being an issue for them. A lot of our customers use third-party tools like cloudify, etc, or heat templates to build instances where multi-instance boot would not be an option as individual calls are made to boot every instance individually so that is probably not the solution. But, given that I haven't heard of this being a major issue, I don't see a problem waiting until Mitaka for the fix.

**Comment by b3rnard0** [comment]
Marking as won't fix.

**Comment by claco** [comment]
The upstream scheduler is being rewritten to address these sorts of problems, and the fix proposed won't solve the issue, only slightly mitigate it in Newton/Mitaka.
